File: Polars.CSharp/Operations/MatchToSchema.cs
using System.Runtime.InteropServices;
using Polars.NET.Core;
using Polars.NET.Core.Native;

namespace Polars.CSharp;

/// <summary>
/// Represents the action to take when a column from the target schema is missing in the current frame.
/// </summary>
public class MissingColumnAction
{
    internal PlMissingColumnsPolicyType Type { get; }
    internal Expr? Expression { get; }

    private MissingColumnAction(PlMissingColumnsPolicyType type, Expr? expr = null)
    {
        Type = type;
        Expression = expr?.Clone();
    }

    /// <summary>
    /// Insert the missing column with null values.
    /// </summary>
    public static MissingColumnAction Insert() => new(PlMissingColumnsPolicyType.Insert);
    
    /// <summary>
    /// Raise an error if the column is missing.
    /// </summary>
    public static MissingColumnAction Raise() => new(PlMissingColumnsPolicyType.Raise);
    
    /// <summary>
    /// Insert the missing column and fill it with the evaluated expression or literal value.
    /// </summary>
    /// <param name="expr">The expression or primitive value to fill the missing column with.</param>
    public static MissingColumnAction InsertWith(IntoExpr expr) 
        => new(PlMissingColumnsPolicyType.InsertWith, expr.Consume());
}

/// <summary>
/// Configuration for schema matching policies during a MatchToSchema operation.
/// Provides fine-grained control over how missing columns, struct fields, and type castings are handled.
/// </summary>
public record MatchSchemaConfig
{
    /// <summary>
    /// Policy for handling top-level columns present in the target schema but missing in the frame.
    /// Defaults to <see cref="MissingColumnAction.Raise"/>.
    /// </summary>
    public MissingColumnAction MissingColumns { get; init; } = MissingColumnAction.Raise();
    
    /// <summary>
    /// Policy for handling missing fields within nested Struct columns.
    /// Defaults to <see cref="MissingColumnsPolicy.Raise"/>.
    /// </summary>
    public MissingColumnsPolicy MissingStructFields { get; init; } = MissingColumnsPolicy.Raise;
    
    /// <summary>
    /// Policy for handling extra fields within nested Struct columns that are not in the target schema.
    /// Defaults to <see cref="ExtraColumnsPolicy.Raise"/>.
    /// </summary>
    public ExtraColumnsPolicy ExtraStructFields { get; init; } = ExtraColumnsPolicy.Raise;
    
    /// <summary>
    /// Policy for handling integer type casting (e.g., allowing upcast from Int32 to Int64).
    /// Defaults to <see cref="UpcastOrForbid.Forbid"/>.
    /// </summary>
    public UpcastOrForbid IntegerCast { get; init; } = UpcastOrForbid.Forbid;
    
    /// <summary>
    /// Policy for handling floating-point type casting (e.g., allowing upcast from Float32 to Float64).
    /// Defaults to <see cref="UpcastOrForbid.Forbid"/>.
    /// </summary>
    public UpcastOrForbid FloatCast { get; init; } = UpcastOrForbid.Forbid;
    internal PolarsWrapper.PlMatchToSchemaConfig ToCoreConfig() => new()
    {
        MissingColumnsType = MissingColumns.Type,
        MissingColumnsExpr = MissingColumns.Expression?.CloneHandle(),
        MissingStructFields = (PlMissingColumnsPolicy)MissingStructFields,
        ExtraStructFields = (PlExtraColumnsPolicy)ExtraStructFields,
        IntegerCast = (PlUpcastOrForbid)IntegerCast,
        FloatCast = (PlUpcastOrForbid)FloatCast
    };
}
public partial class LazyFrame : IDisposable,IPolarsLazyFrame
{
    /// <summary>
    /// Align the Frame's schema to the target schema.
    /// Columns not in the target schema are handled according to the extraColumns policy.
    /// </summary>
    /// <param name="schema">The target schema to match against.</param>
    /// <param name="extraColumns">Policy for handling top-level columns present in the frame but not in the target schema.</param>
    /// <param name="defaultConfig">The default configuration applied to all columns for handling missing data and type casting.</param>
    /// <param name="columnOverrides">Specific configuration overrides mapped by column name.</param>
    /// <returns>A new <see cref="LazyFrame"/> with the aligned schema.</returns>
    /// <exception cref="ArgumentNullException">Thrown when the provided schema is null.</exception>

    public LazyFrame MatchToSchema(
        PolarsSchema schema,
        ExtraColumnsPolicy extraColumns = ExtraColumnsPolicy.Raise,
        MatchSchemaConfig? defaultConfig = null,
        IReadOnlyDictionary<string, MatchSchemaConfig>? columnOverrides = null)
    {
        ArgumentNullException.ThrowIfNull(schema);

        var coreDefault = (defaultConfig ?? new MatchSchemaConfig()).ToCoreConfig();

        Dictionary<string,PolarsWrapper.PlMatchToSchemaConfig>? coreOverrides = null;
        if (columnOverrides != null && columnOverrides.Count > 0)
        {
            coreOverrides = new Dictionary<string, PolarsWrapper.PlMatchToSchemaConfig>(columnOverrides.Count);
            foreach (var kvp in columnOverrides)
            {
                coreOverrides[kvp.Key] = kvp.Value.ToCoreConfig();
            }
        }

        var handle = PolarsWrapper.MatchToSchema(
            this.CloneHandle(),
            schema.Handle,
            (PlExtraColumnsPolicy)extraColumns,
            coreDefault,
            coreOverrides
        );

        return new LazyFrame(handle);
    }
}

public partial class DataFrame : IDisposable,IEnumerable<Series>,IPolarsDataFrame
{
    /// <inheritdoc cref="LazyFrame.MatchToSchema(PolarsSchema, ExtraColumnsPolicy, MatchSchemaConfig?, IReadOnlyDictionary{string, MatchSchemaConfig}?)"/>
    public DataFrame MatchToSchema(
        PolarsSchema schema,
        ExtraColumnsPolicy extraColumns = ExtraColumnsPolicy.Raise,
        MatchSchemaConfig? defaultConfig = null,
        IReadOnlyDictionary<string, MatchSchemaConfig>? columnOverrides = null)
    => Lazy().MatchToSchema(schema,extraColumns,defaultConfig,columnOverrides).Collect();
}

File: Polars.FSharp/Operations/MatchToSchema.fs
namespace Polars.FSharp

open Polars.NET.Core
open System

/// <summary>
/// Represents the action to take when a column from the target schema is missing in the current frame.
/// </summary>
type MissingColumnAction private (policyType: PlMissingColumnsPolicyType, expr: Expr option) =
    member internal this.Type = policyType
    member internal this.Expression = expr

    /// <summary>
    /// Insert the missing column with null values.
    /// </summary>
    static member Insert() = MissingColumnAction(PlMissingColumnsPolicyType.Insert, None)

    /// <summary>
    /// Raise an error if the column is missing.
    /// </summary>
    static member Raise() = MissingColumnAction(PlMissingColumnsPolicyType.Raise, None)

    /// <summary>
    /// Insert the missing column and fill it with the evaluated expression or literal value.
    /// </summary>
    /// <param name="expr">The expression to fill the missing column with.</param>
    static member InsertWith(expr: Expr) = 
        MissingColumnAction(PlMissingColumnsPolicyType.InsertWith, Some (expr.Clone()))

/// <summary>
/// Configuration for schema matching policies during a MatchToSchema operation.
/// Provides fine-grained control over how missing columns, struct fields, and type castings are handled.
/// </summary>
type MatchSchemaConfig =
    { /// Policy for handling top-level columns present in the target schema but missing in the frame.
      MissingColumns: MissingColumnAction
      /// Policy for handling missing fields within nested Struct columns.
      MissingStructFields: MissingColumnsPolicy
      /// Policy for handling extra fields within nested Struct columns that are not in the target schema.
      ExtraStructFields: ExtraColumnsPolicy
      /// Policy for handling integer type casting.
      IntegerCast: UpcastOrForbid
      /// Policy for handling floating-point type casting.
      FloatCast: UpcastOrForbid }

    /// <summary>
    /// Creates a default schema matching configuration.
    /// </summary>
    static member Default =
        { MissingColumns = MissingColumnAction.Raise()
          MissingStructFields = MissingColumnsPolicy.Raise
          ExtraStructFields = ExtraColumnsPolicy.Raise
          IntegerCast = UpcastOrForbid.Forbid
          FloatCast = UpcastOrForbid.Forbid }

    member internal this.ToCoreConfig() : PolarsWrapper.PlMatchToSchemaConfig =
        let mutable c = PolarsWrapper.PlMatchToSchemaConfig()
        c.MissingColumnsType <- this.MissingColumns.Type
        c.MissingColumnsExpr <- 
            match this.MissingColumns.Expression with
            | Some e -> e.CloneHandle()
            | None -> null
        c.MissingStructFields <- this.MissingStructFields.ToNative()
        c.ExtraStructFields <- this.ExtraStructFields.ToNative()
        c.IntegerCast <- this.IntegerCast.ToNative()
        c.FloatCast <- this.FloatCast.ToNative()
        c


[<AutoOpen>]
module MatchToSchemaOps =
    open System.Collections.Generic
    type LazyFrame with
        member this.MatchToSchema(schema: PolarsSchema, 
                                ?extraColumns: ExtraColumnsPolicy,
                                ?defaultConfig: MatchSchemaConfig,
                                ?columnOverrides: Map<string, MatchSchemaConfig>) : LazyFrame =
                                
            let exa = defaultArg extraColumns ExtraColumnsPolicy.Raise
            let coreDefault = (defaultArg defaultConfig MatchSchemaConfig.Default).ToCoreConfig()

            let coreOverrides = 
                columnOverrides
                |> Option.filter (fun m -> not m.IsEmpty)
                |> Option.map (fun m ->
                    let d = Dictionary<_,_>()
                    m |> Map.iter (fun k v -> d.[k] <- v.ToCoreConfig())
                    d
                )
                |> Option.toObj

            let handle = 
                PolarsWrapper.MatchToSchema(
                    this.CloneHandle(),
                    schema.Handle,
                    exa.ToNative(),
                    coreDefault,
                    coreOverrides
                )

            new LazyFrame(handle)